openssl/goopenssl.c

```diff
@@ -9,6 +9,27 @@
 #include <dlfcn.h>
 #include <stdio.h>
 
+int
+go_openssl_fips_enabled(void* handle)
+{
+    // For OpenSSL 1.x.
+    int (*FIPS_mode)(void);
+    FIPS_mode = (int (*)(void))dlsym(handle, "FIPS_mode");
+    if (FIPS_mode != NULL)
+        return FIPS_mode();
```

Comment:
Hmm, on second look, I think this might be going too deep? If systemwide/opensslwide FIPS isn't enabled, it seems like this means that even if FIPS mode is *supported*, this will return `0`, so this version of OpenSSL won't get picked. So if we e.g. have 3.0 that doesn't support FIPS and 1.0.2 that supports FIPS but FIPS isn't enabled by default, we would pick 3.0 as the "fallback", not have FIPS, and (e.g.) fail to turn on FIPS mode later when we really should have been able to.

Reply:
Yes, that was may intention, pick the systemwide enabled OpenSSL version, not just one that is FIPS-capable but not enabled. IMO it makes more sense this way, production environments that require FIPS must have it enable systemwide, as not all tooling and framework provides something like `GOFIPS=1`.